vxquery-core/src/main/java/org/apache/vxquery/jsonparser/JSONParser.java

```diff
@@ -0,0 +1,184 @@
+ * the License. You may obtain a copy of the License at
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.vxquery.jsonparser;
+
+import java.io.DataOutput;
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.apache.htrace.fasterxml.jackson.core.JsonFactory;
+import org.apache.htrace.fasterxml.jackson.core.JsonParseException;
+import org.apache.htrace.fasterxml.jackson.core.JsonParser;
+import org.apache.htrace.fasterxml.jackson.core.JsonToken;
+import org.apache.hyracks.data.std.primitive.UTF8StringPointable;
+import org.apache.hyracks.data.std.util.ArrayBackedValueStorage;
+import org.apache.vxquery.datamodel.builders.atomic.StringValueBuilder;
+import org.apache.vxquery.datamodel.builders.jsonitem.ArrayBuilder;
+import org.apache.vxquery.datamodel.builders.jsonitem.ObjectBuilder;
+import org.apache.vxquery.datamodel.values.ValueTag;
+import org.apache.vxquery.xmlparser.IParser;
+
+public class JSONParser implements IParser{
+
+    final JsonFactory factory;
+    protected final ArrayBackedValueStorage atomic;
+    protected final List<ArrayBuilder> abStack;
+    protected final List<ObjectBuilder> obStack;
+    protected final List<ArrayBackedValueStorage> abvsStack;
+    protected final List<ArrayBackedValueStorage> keyStack;
+    protected final List<UTF8StringPointable> spStack;
+    protected final StringValueBuilder svb;
+
+    enum itemType {
+        ARRAY,
+        OBJECT
+    }
+
+    protected final List<itemType> itemStack;
+
+    public JSONParser() throws JsonParseException {
+        factory = new JsonFactory();
+        atomic = new ArrayBackedValueStorage();
+        abStack = new ArrayList<ArrayBuilder>();
+        obStack = new ArrayList<ObjectBuilder>();
+        abvsStack = new ArrayList<ArrayBackedValueStorage>();
+        keyStack = new ArrayList<ArrayBackedValueStorage>();
+        spStack = new ArrayList<UTF8StringPointable>();
+        itemStack = new ArrayList<itemType>();
+        svb = new StringValueBuilder();
+    }
+
+    public void parseDocument(File file, ArrayBackedValueStorage result)
+            throws NumberFormatException, JsonParseException, IOException {
+        DataOutput outResult = result.getDataOutput();
+        abvsStack.add(atomic);
+        DataOutput out = abvsStack.get(0).getDataOutput();
+        String startItem = null;
+        JsonParser parser = factory.createParser(file);
+        JsonToken token = parser.nextToken();
+        itemType checkItem = null;
+        int levelArray = 0;
+        int levelObject = 0;
```

Comment:
Take a close look at the parseDocument method and move all variable that can be initialized and/or declared out of this method. You can reset them, instead of creating a new variable.